File: Graph/133. Clone Graph/cloneGraph.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val = 0, neighbors = None):
        self.val = val
        self.neighbors = neighbors if neighbors is not None else []
"""

class Solution:
    def cloneGraph(self, node: 'Node', map = {}) -> 'Node':
        if not node:
            return None
        
        if node in map:
            return map[node]
        
        # copy is the cloned node
        # it is an entry point of all nodes and neighbors
        copy = Node(node.val)
        # the only function of map is to record visited node
        map[node] = copy
        
        for n in node.neighbors:
            # update copy's neighbors 
            copy.neighbors.append(self.cloneGraph(n, map))
            
        return copy


"""
# Definition for a Node.
class Node:
    def __init__(self, val = 0, neighbors = None):
        self.val = val
        self.neighbors = neighbors if neighbors is not None else []
"""

class Solution:
    def cloneGraph(self, node: 'Node', map = {}) -> 'Node':
        if not node:
            return None
        
        map[node] = Node(node.val)
        queue = []
        queue.append(node)
        
        while len(queue):
            current = queue.pop(0)
            for n in current.neighbors:
                if n not in map:
                    # record a new node
                    map[n] = Node(n.val)
                    # add to queue
                    queue.append(n)
                # two nodes connect with each other
                map[n].neighbors.append(map.get(current))
                
        return map[node]

File: Graph/133. Clone Graph/cloneGraph.js
/**
 * // Definition for a Node.
 * function Node(val, neighbors) {
 *    this.val = val === undefined ? 0 : val;
 *    this.neighbors = neighbors === undefined ? [] : neighbors;
 * };
 */
/**
 * @param {Node} node
 * @return {Node}
 */

// DFS
 var cloneGraph = function(node, map = new Map()) {
    if (!node) return null;
    
    // O(n) as map is to memorise the visited node
    if (map.has(node)) 
        return map.get(node);
    
    // cloned node
    let copy = new Node(node.val);
    // map is to record the visited node
    map.set(node, copy);
    // traverse node's neighbors
    for (let n of node.neighbors) {
        copy.neighbors.push(cloneGraph(n, map));
    }
    
    return copy;
};


 // BFS
 var cloneGraph = function(node, map = new Map()) {
    if (!node) return null;
    
    map.set(node, new Node(node.val));
    let queue = [node];
    
    while (queue.length) {
        let current = queue.shift();
        for (let n of current.neighbors) {
            if (!map.has(n)) {
                map.set(n, new Node(n.val));
                queue.push(n);
            } 
            //undirected so both edges need to be connected
            map.get(n).neighbors.push(map.get(current));
        }
    }
    
    return map.get(node);
};

